app/graphql/mutations/settings/update.rb

```diff
@@ -1,24 +1,59 @@
 module Mutations
   module Settings
-    class Update < UpdateMutation
+    class Update < BaseMutation
       graphql_name 'UpdateSettingMutation'
       description 'Updates a setting'
 
       # https://github.com/graphql/graphql-spec/blob/master/rfcs/InputUnion.md
+      argument :id, ID, required: false
+      argument :name, String, required: false
       argument :value, String, required: true
 
-      field :setting, Types::Setting, 'Setting', null: true
+      field :errors, [Types::AttributeError], null: false
+      field :setting, Types::Setting, description: 'Setting', null: true
 
-      def assign_attributes(object, params)
-        object.parse_string_value(params[:value])
+      def self.resource_class
+        SettingPresenter
       end
 
-      def self.save_object(resource)
-        return super resource unless resource.errors.any?
-        {
-          result_key => resource,
-          :errors => map_errors_to_path(resource),
-        }
+      def resolve(params)
+        authorize!
+
+        if params[:name]
+          name = params[:name]
+        else
+          _, _, name = Foreman::GlobalId.decode(params[:id])
+        end
+
+        definition = Foreman.settings.find(name)
+        validate_object(definition)
+
+        record = Foreman.settings.set_user_value(definition.name, params[:value])
+        save_object(definition, record)
+      end
+
+      def save_object(definition, record)
+        User.as(context[:current_user]) do
+          errors = if record.save
+                     []
+                   else
+                     map_errors_to_path(record)
+                   end
+          {
+            :setting => definition,
+            :errors => errors,
+          }
+        end
+      end
+
+      private
+
+      def authorize!
+        return if context[:current_user].can?(:edit_settings)
```

Comment:
`context[:current_user]` can be `nil`, leading to

```
2022-02-16T11:29:17 [I|app|018c699d] Completed 500 Internal Server Error in 3ms (ActiveRecord: 0.0ms | Allocations: 2463)
 018c699d | 
 018c699d | 
2022-02-16T11:29:17 [F|app|018c699d]   
 018c699d | NoMethodError (undefined method `can?' for nil:NilClass):
 018c699d |   
 018c699d | app/graphql/mutations/settings/update.rb:52:in `authorize!'
 018c699d | app/graphql/mutations/settings/update.rb:20:in `resolve'
 018c699d | app/controllers/api/graphql_controller.rb:60:in `execute_single_graphql_query'
 018c699d | app/controllers/api/graphql_controller.rb:21:in `execute'
```

Reply:
What situations this would be nil? Is it safe to assume that in such cases the user is not authorized?

Reply:
It happened when I tried sending requests from graphiql without logging into foreman first.

> Is it safe to assume that in such cases the user is not authorized?

I'd say yes